File: src/runtime-core/exchange-alert-routing.ts
export interface ExchangeAlertThread {
  waitingOn?: string[];
}

export interface ExchangeAlertMessage {
  deliveredTo?: string[];
  to?: string[];
}

export interface ExchangeAlertTask {
  assignedBy?: string;
  assignedTo?: string;
}

function uniqueTrimmed(values: Array<string | undefined>): string[] {
  return [...new Set(values.map((value) => value?.trim()).filter(Boolean) as string[])];
}

function recipientsExcludingSender(values: Array<string | undefined>, from?: string): string[] {
  const sender = from?.trim();
  return uniqueTrimmed(values).filter((recipient) => recipient !== sender);
}

function stringArrayField(value: unknown, field: "deliveredTo" | "to" | "waitingOn"): string[] {
  if (!value || typeof value !== "object") return [];
  const fieldValue = (value as Record<string, unknown>)[field];
  return Array.isArray(fieldValue) ? fieldValue.filter((item): item is string => typeof item === "string") : [];
}

export function resolveExchangeMessageAlertRecipients(input: {
  explicitRecipients?: string[];
  message?: unknown;
  thread?: unknown;
  fallbackRecipients?: string[];
  from?: string;
}): string[] {
  const explicit = uniqueTrimmed(input.explicitRecipients ?? []);
  if (explicit.length > 0) return recipientsExcludingSender(explicit, input.from);

  const deliveredTo = uniqueTrimmed(stringArrayField(input.message, "deliveredTo"));
  if (deliveredTo.length > 0) return recipientsExcludingSender(deliveredTo, input.from);

  const waitingOn = uniqueTrimmed(stringArrayField(input.thread, "waitingOn"));
  if (waitingOn.length > 0) return recipientsExcludingSender(waitingOn, input.from);

  const messageRecipients = uniqueTrimmed(stringArrayField(input.message, "to"));
  if (messageRecipients.length > 0) return recipientsExcludingSender(messageRecipients, input.from);

  return recipientsExcludingSender(input.fallbackRecipients ?? [], input.from);
}

export function resolveExchangeTaskAssignmentRecipient(task: ExchangeAlertTask): string | undefined {
  return uniqueTrimmed([task.assignedTo])[0];
}

export function resolveExchangeTaskOutcomeRecipient(input: {
  task: ExchangeAlertTask;
  thread?: unknown;
  from?: string;
}): string | undefined {
  const waitingOn = uniqueTrimmed(stringArrayField(input.thread, "waitingOn"));
  if (waitingOn.length > 0) return recipientsExcludingSender(waitingOn, input.from)[0];
  return recipientsExcludingSender([input.task.assignedBy], input.from)[0];
}

export function resolveExchangeReviewOutcomeRecipient(task: ExchangeAlertTask): string | undefined {
  return uniqueTrimmed([task.assignedBy])[0];
}

File: native/crates/aimux/src/exchange_alert_routing.rs
use std::collections::BTreeSet;

use serde_json::Value;

pub fn resolve_exchange_alert_routing(case: &Value) -> Value {
    match case["api"].as_str().unwrap_or_default() {
        "resolveExchangeMessageAlertRecipients" => {
            string_array(resolve_exchange_message_alert_recipients(&case["input"]))
        }
        "resolveExchangeTaskAssignmentRecipient" => optional_string(
            resolve_exchange_task_assignment_recipient(&case["input"])
                .as_deref(),
        ),
        "resolveExchangeTaskOutcomeRecipient" => {
            optional_string(resolve_exchange_task_outcome_recipient(&case["input"]).as_deref())
        }
        "resolveExchangeReviewOutcomeRecipient" => optional_string(
            resolve_exchange_review_outcome_recipient(&case["input"])
                .as_deref(),
        ),
        _ => Value::Null,
    }
}

pub fn resolve_exchange_message_alert_recipients(input: &Value) -> Vec<String> {
    let explicit = unique_trimmed_array(input.get("explicitRecipients"));
    if !explicit.is_empty() {
        return recipients_excluding_sender(explicit, string_field(input, "from").as_deref());
    }

    let delivered_to = unique_trimmed_array(input.get("message").and_then(|message| message.get("deliveredTo")));
    if !delivered_to.is_empty() {
        return recipients_excluding_sender(delivered_to, string_field(input, "from").as_deref());
    }

    let waiting_on = unique_trimmed_array(input.get("thread").and_then(|thread| thread.get("waitingOn")));
    if !waiting_on.is_empty() {
        return recipients_excluding_sender(waiting_on, string_field(input, "from").as_deref());
    }

    let message_recipients = unique_trimmed_array(input.get("message").and_then(|message| message.get("to")));
    if !message_recipients.is_empty() {
        return recipients_excluding_sender(message_recipients, string_field(input, "from").as_deref());
    }

    recipients_excluding_sender(
        unique_trimmed_array(input.get("fallbackRecipients")),
        string_field(input, "from").as_deref(),
    )
}

pub fn resolve_exchange_task_assignment_recipient(task: &Value) -> Option<String> {
    unique_trimmed([string_field(task, "assignedTo")].into_iter()).into_iter().next()
}

pub fn resolve_exchange_task_outcome_recipient(input: &Value) -> Option<String> {
    let waiting_on = unique_trimmed_array(input.get("thread").and_then(|thread| thread.get("waitingOn")));
    if !waiting_on.is_empty() {
        return recipients_excluding_sender(waiting_on, string_field(input, "from").as_deref())
            .into_iter()
            .next();
    }
    recipients_excluding_sender(
        unique_trimmed([input.get("task").and_then(|task| string_field(task, "assignedBy"))].into_iter()),
        string_field(input, "from").as_deref(),
    )
    .into_iter()
    .next()
}

pub fn resolve_exchange_review_outcome_recipient(task: &Value) -> Option<String> {
    unique_trimmed([string_field(task, "assignedBy")].into_iter()).into_iter().next()
}

fn recipients_excluding_sender(values: Vec<String>, from: Option<&str>) -> Vec<String> {
    let sender = from.map(str::trim).filter(|value| !value.is_empty());
    values
        .into_iter()
        .filter(|recipient| Some(recipient.as_str()) != sender)
        .collect()
}

fn unique_trimmed_array(value: Option<&Value>) -> Vec<String> {
    unique_trimmed(
        value
            .and_then(Value::as_array)
            .into_iter()
            .flatten()
            .map(|item| item.as_str().map(str::to_owned)),
    )
}

fn unique_trimmed(values: impl Iterator<Item = Option<String>>) -> Vec<String> {
    let mut seen = BTreeSet::new();
    let mut unique = Vec::new();
    for value in values {
        let Some(value) = value.map(|value| value.trim().to_owned()) else {
            continue;
        };
        if value.is_empty() || !seen.insert(value.clone()) {
            continue;
        }
        unique.push(value);
    }
    unique
}

fn string_array(values: Vec<String>) -> Value {
    Value::Array(values.into_iter().map(Value::String).collect())
}

fn optional_string(value: Option<&str>) -> Value {
    value.map(|value| Value::String(value.into())).unwrap_or(Value::Null)
}

fn string_field(value: &Value, key: &str) -> Option<String> {
    value.get(key).and_then(Value::as_str).map(str::to_owned)
}
